src/bridge/DelayBuffer.sol

```diff
@@ -0,0 +1,177 @@
+    /// @param start The beginning reference point (prev delay)
+    /// @param end The ending reference point (current message)
+    /// @param prevDelay The delay to be applied
+    /// @param threshold The threshold to saturate at
+    /// @param buffer The buffer to be decremented
+    function deplete(
+        uint256 start,
+        uint256 end,
+        uint256 buffer,
+        uint256 prevDelay,
+        uint256 threshold
+    ) internal pure returns (uint256) {
+        uint256 elapsed = end > start ? end - start : 0;
+        uint256 unexpectedDelay = prevDelay > threshold ? prevDelay - threshold : 0;
+        if (unexpectedDelay > elapsed) {
+            unexpectedDelay = elapsed;
+        }
+        // decrease the buffer saturating at zero
+        buffer = buffer > unexpectedDelay ? buffer - unexpectedDelay : 0;
+        // saturate at threshold
+        buffer = buffer > threshold ? buffer : threshold;
+        return buffer;
+    }
+
+    /// @notice Replenishes the delay buffer saturating at max
+    /// @param start The beginning reference point
+    /// @param end The ending reference point
+    /// @param buffer The buffer to be replenished
+    /// @param max The maximum buffer
+    /// @param replenishRateInBasis The amount to replenish the buffer per block in basis points.
+    function replenish(
+        uint256 start,
+        uint256 end,
+        uint256 buffer,
+        uint256 max,
+        uint256 replenishRateInBasis
+    ) internal pure returns (uint256) {
+        // add the replenish round off from the last replenish
+        uint256 elapsed = end > start ? end - start : 0;
+        // rounds down for simplicity
+        uint256 replenishAmount = (elapsed * replenishRateInBasis) / BASIS;
+        if (max - buffer > replenishAmount) {
+            buffer += replenishAmount;
+        } else {
+            // saturate
+            buffer = max;
+        }
+        return buffer;
+    }
+
+    /// @notice Conditionally updates the buffer. Depletes if the delay is unexpected, otherwise replenishes if the buffer is depleted.
+    /// @param start The beginning reference point
+    /// @param end The ending reference point
+    /// @param buffer The buffer to be replenished
+    /// @param max The maximum buffer
+    /// @param replenishRateInBasis The amount to replenish the buffer per block in basis points.
+    function bufferUpdate(
+        uint256 start,
+        uint256 end,
+        uint256 buffer,
+        uint256 prevDelay,
+        uint256 threshold,
+        uint256 max,
+        uint256 replenishRateInBasis
+    ) internal pure returns (uint256) {
+        if (threshold < prevDelay) {
+            // deplete due to unexpected delay
+            buffer = deplete(start, end, buffer, prevDelay, threshold);
+        } else if (buffer < max) {
+            // replenish depleted buffer
+            buffer = replenish(start, end, buffer, max, replenishRateInBasis);
+        }
```

Comment:
while in delay mode (i.e. `threshold < prevDelay`), the sequencer would forgo all buffer replenishment
e.g. if the sequencer delayed `threshold + 1`, not only he would deplete 1 second of buffer, he also would not be able to get the `threshold * rate` worth of replenishment

I am ok with the implementation tho we can also always replenish so that it might be easier to specify the sequencer always have `rate` per second of delay buffer

Reply:
always replenish is a simpler policy, I went ahead and simplified the buffer update logic to always replenish.